api/src/middleware/cache_resource.rs

```
@@ -56,25 +72,39 @@ impl Middleware<AppState> for CacheResource {
             let state = request.state().clone();
             let config = state.config.clone();
-            let mut user_value = "".to_string();
-            match OptionalUser::from_request(request, &()) {
-                Ok(user) => {
-                    cache_configuration.user_id = user.0.map(|u| u.id());
-                    if let Some(user_id) = cache_configuration.user_id {
-                        user_value = user_id.to_string();
+
+            if self.cache_users_by != CacheUsersBy::None {
+                let user = match OptionalUser::from_request(request, &()) {
+                    Ok(user) => user,
+                    Err(error) => {
+                        cache_configuration.error = true;
+                        error!("CacheResource Middleware start: {:?}", error);
+                        request.extensions_mut().insert(cache_configuration);
+                        return Ok(Started::Done);
+                    }
+                };
+                if let Some(user) = user.0 {
+                    match self.cache_users_by {
+                        CacheUsersBy::None => (),
+                        CacheUsersBy::AnonymousOnly => {
+                            // Do not cache
+                            return Ok(Started::Done);
+                        }
+                        CacheUsersBy::UserId => {
+                            cache_configuration.user_key = Some(user.id().to_string());
+                        }
+                        CacheUsersBy::GlobalRoles => {
+                            cache_configuration.user_key = Some(user.user.role.iter().map(|r| r.to_string()).join(","));
+                        }
+                    }
+                    if let Some(ref user_key) = cache_configuration.user_key {
+                        query.insert(&user_text, user_key.to_string());
                     }
-                    query.insert(&user_text, &user_value);
-                }
-                Err(error) => {
```

My thinking for the error here was that given we're swallowing the error here in the event it failed here but succeeded in the controller in all of the other requests we could end up caching user specific data as the non user cache. It's an extreme edge case though.